Replace existing breaking changes documentation

conniey opened the pull request:
# Replace existing breaking changes documentation

## Summary

* Replaces existing [application compatibility documentation](https://docs.microsoft.com/en-us/dotnet/articles/framework/migration-guide/application-compatibility) with ones generated from Microsoft/dotnet.
 * This should replace all those different links like [Application Compatibility in 4.5](https://msdn.microsoft.com/en-us/library/hh367887(v=vs.110).aspx), [Application Compatibility in 4.5.1](https://msdn.microsoft.com/en-us/library/hh367887(v=vs.110).aspx), etc. with the use of the multi-selectors. 
* Generated from [microsoft/dotnet](https://github.com/Microsoft/dotnet/tree/master/Documentation/breaking-changes)
   * If there needs to be wording changes for a particular breaking change, I'll also have to make a PR in  Microsoft/dotnet since these are auto-generated.

## Suggested Reviewers

/cc: @mairaw , @rpetrusha , @twsouthwick

conniey committed:
Remove unnecessary retargeting/runtime change files

rpetrusha commented on docs/framework/migration-guide/application-compatibility.md:
```diff
@@ -15,128 +15,81 @@ helpviewer_keywords:
+assembly, so the console app that uses the assembly, or the website that
+references the assembly.
+
+The .NET Compatibility Diagnostics are Roslyn-powered analyzers that help
```
Add another header here: ## .NET Compatibility Diagnostics

rpetrusha commented on docs/framework/migration-guide/application-compatibility.md:
```diff
@@ -15,128 +15,81 @@ helpviewer_keywords:
+
+-   The description of what has changed from a previous version.
+
+-   The suggestion is a description of how the change affects customers and whether any workarounds are available to preserve compatibility across versions.
```
Just "How the change affects customers..."

rpetrusha commented on docs/framework/migration-guide/application-compatibility.md:
```diff
@@ -15,128 +15,81 @@ helpviewer_keywords:
+    |Edge case|A change that affects apps under very specific, uncommon scenarios.|
+    |Transparent|A change with no noticeable effect on the application's developer or user.|
+
+-   Version indicates when the change first appears in the framework. Some of the changes are reverted and that is indicated as well.
```
"reverted" is likely to cause confusion. "Some of the changes are introduced in a particular version and reverted in a later version; that is indicated as well."

rpetrusha commented on docs/framework/migration-guide/application-compatibility.md:
```diff
@@ -15,128 +15,81 @@ helpviewer_keywords:
+    |   |   |
+    |---|---|
+    |Retargeting|The change affects apps that are recompiled to target a new version of the .NET Framework.|
+    |Runtime|The change affects an existing app that targets a previous version of the .NET Framework but runs on a later version.|
```
Runtime changes affect all apps running on a particular version; targeting is not relevant.

rpetrusha commented on docs/framework/migration-guide/retargeting/4.0-4.5.1.md:
```diff
@@ -0,0 +1,325 @@
+
+* Changes in the runtime environment. These affect only apps that specifically target the retargeted .NET Framework. Apps that target previous versions of the .NET Framework behave as they did when running under those versions.
+
+In the topics that describe etargeting changes, we have classified individual items by their expected impact, as follows:
```
etargeting -> retargeting

rpetrusha commented on docs/framework/migration-guide/retargeting/4.0-4.5.1.md:
```diff
@@ -0,0 +1,325 @@
+
+|   |   |
+|---|---|
+|Details|DbParameter.Precision and DbParameter.Scale are implemented as public virtual properties. They replace the corresponding explicit interface implementations, DbParameter.IDbDataParameter.Precision and DbParameter.IDbDataParameter.Scale.|
```
Preferably, these should all be links, especially the EIIs.

rpetrusha commented on docs/framework/migration-guide/retargeting/4.0-4.5.1.md:
```diff
@@ -0,0 +1,325 @@
+|   |   |
+|---|---|
+|Details|These methods are now obsolete. Compilation of code that calls these methods produces a compiler warning.|
+|Suggestion|The recommended alternatives are <a href="https://msdn.microsoft.com/en-us/library/system.web.security.machinekey.protect(v=vs.110).aspx">MachineKey.Protect</a> and <a href="https://msdn.microsoft.com/en-us/library/system.web.security.machinekey.unprotect(v=vs.110).aspx">MachineKey.Unprotect</a>. Alternatively, the build warnings can be suppressed or they can be avoided by using an ol ...
```
Use <xref>s to docs rather than links to MSDN.

rpetrusha commented on docs/framework/migration-guide/retargeting/4.0-4.5.1.md:
```diff
@@ -0,0 +1,325 @@
+|   |   |
+|---|---|
+|Details|For applications that target the .NET Framework 4.5, characters that are outside the Basic Multilingual Plane (BMP) round-trip correctly when they are passed to the WebUtility.HtmlDecode method.|
+|Suggestion|This change should have no effect on current applications, but to restore the original behavior, set the targetFramework attribute of the <code>&lt;httpRuntime&gt;</code> element to a string other than &quot;4.5&quot;. You can also set the <code>unicodeEncodingConformance</code> and <code>unicodeDecodingConformance</code> attributes of the <code>&lt;webUtility&gt;</code> configuration ...
```
`targetFramework` or <code>targetFramework</code>

rpetrusha commented on docs/framework/migration-guide/retargeting/4.0-4.5.1.md:
```diff
@@ -0,0 +1,325 @@
+
+|   |   |
+|---|---|
+|Details|Beginning in .NET 4.5, a <code>List&lt;T&gt;.ForEach</code> enumerator will throw an InvalidOperationException exception if an element in the calling collection is modified. Previously, this would not throw an exception but could lead to race conditions.|
```
link to InvalidOperationException

rpetrusha commented on docs/framework/migration-guide/retargeting/4.0-4.5.1.md:
```diff
@@ -0,0 +1,325 @@
+
+## MSBuild
+
+### 49: ResolveAssemblyReference task now warns on dependencies with the wrong architecture
```
warns on --> warns of

conniey replied:
Fixed.

conniey commented on docs/framework/migration-guide/application-compatibility.md:
```diff
@@ -15,128 +15,81 @@ helpviewer_keywords:
+on the older version. Many of the breaking changes between versions of the .NET
+Framework are mitigated through this quirking model.
+
+Runtime issues are those that arise when a new runtime is placed on a machine
```
I think the paragraph describes runtime issues rather than retargeting issues.  Runtime issues are differences that are seen when a user installs a newer version of .NET Framework on their machine and then runs their application (which is targeting an older version).  Retargeting changes are differences during runtime when users recompile their code against a newer version of the .NET Framework.

From [Runtime Changes in the .NET Framework 4.5.1](https://msdn.microsoft.com/en-us/library/dn458360(v=vs.110).aspx)
> ... runtime changes may affect existing apps that target the .NET Framework 4 or 4.5 but run on the 4.51 runtime.

From [Retargeting Changes in the .NET Framework 4.5.1](https://msdn.microsoft.com/en-us/library/dn458356(v=vs.110).aspx)
> ...retargeting changes may affect apps that are recompiled to target the .NET Framework 4.5.1. 


@twsouthwick :

rpetrusha commented:
Is this ready to merge, @conniey?

conniey commented:
@rpetrusha : No. It's not. I want to make some changes to the files structure to reduce the duplication of breaking change content. I'll ping you again when it is.

rpetrusha commented:
OK. Thanks, @conniey.

mairaw commented:
Labeled as WIP then so we know it's ready to merge

conniey committed:
Clean up Table of Contents

conniey commented:
@rpetrusha @mairaw I've fixed the PR. Here is that content now: https://review.docs.microsoft.com/en-us/dotnet-new/articles/framework/migration-guide/application-compatibility?branch=pr-en-us-1923

mairaw commented:
I'll review this now but I've noticed two things right away:
1 - the build report has some warnings about the new topics. Status should say passed instead of warnings (other warnings will exist for other articles)
2 - I've noticed you're removing some articles. When we do that, we need to redirect the content via this file: https://github.com/dotnet/docs/blob/master/.openpublishing.redirection.json. Otherwise it can cause SEO issues plus existing external links need to know where to go.

rpetrusha commented:
@conniey, in looking at the build warnings, it looks as if a significant number of build issues are caused by removing migration-guide/index.md. Did you really mean to remove it? It should remain the top-level migration page, and correspond roughly to migration-guide/toc.md.

mairaw commented on docs/framework/migration-guide/application-compatibility.md:
```diff
@@ -1,142 +1,110 @@
+
+Runtime issues are those that arise when a new runtime is placed on a machine
+and the same binaries are run, but different behavior is seen. If a binary was
+compiled for .NET 4.0 it will run in .NET 4.0 compatibility mode on 4.5+. Many
```
.NET -> .NET Framework
4.5+ -> couldn't find a definitive answer on the style guide about this but I don't think 4.5+ localizes well. My recommendation would be to change this to 4.5 or later versions, for example.

conniey committed:
Creating Breaking Changes document from microsoft\dotnet

conniey committed:
Updating .openpublishing.redirection for pages removed

mairaw reviewed (changes requested):
It's a large PR so I'll send chunks of comments at a time

mairaw commented on docs/framework/migration-guide/retargeting/dbparameterprecision_and_dbparameterscale_are_now_public_virtual_members.md:
```diff
@@ -0,0 +1,11 @@
+
+|   |   |
+|---|---|
+|Details|<xref:System.Data.Common.DbParameter.Precision> and <xref:System.Data.Common.DbParameter.Scale> are implemented as public virtual properties. They replace the corresponding explicit interface implementations, <xref:System.Data.Common.DbParameter.System#Data#IDbDataParameter#Precision> and <xref:System.Data.Common.DbParameter.System#Data#IDbDataParameter#Scale>.|
```
`#` needs to be encoded in xref links. For more info, see https://github.com/dotnet/docs/blob/master/styleguide/template.md#links-to-apis

mairaw commented on docs/framework/migration-guide/retargeting/dbparameterprecision_and_dbparameterscale_are_now_public_virtual_members.md:
```diff
@@ -0,0 +1,11 @@
+|   |   |
+|---|---|
+|Details|<xref:System.Data.Common.DbParameter.Precision> and <xref:System.Data.Common.DbParameter.Scale> are implemented as public virtual properties. They replace the corresponding explicit interface implementations, <xref:System.Data.Common.DbParameter.System#Data#IDbDataParameter#Precision> and <xref:System.Data.Common.DbParameter.System#Data#IDbDataParameter#Scale>.|
+|Suggestion|When re-building an ADO.NET database provider, these differences will require the &#39;override&#39; keyword to be applied to the Precision and Scale properties. This is only needed when re-building the components; exisiting binaries will continue to work.|
```
typo: exisiting

mairaw commented on docs/framework/migration-guide/retargeting/machinekeyencode_and_machinekeydecode_methods_are_now_obsolete.md:
```diff
@@ -0,0 +1,11 @@
+|   |   |
+|---|---|
+|Details|These methods are now obsolete. Compilation of code that calls these methods produces a compiler warning.|
+|Suggestion|The recommended alternatives are <xref:System.Web.Security.MachineKey.Protect(System.Byte[],System.String[])> and <xref:System.Web.Security.MachineKey.Unprotect(System.Byte[],System.String[])>. Alternatively, the build warnings can be suppressed or they can be avoided by using an older compiler. The APIs are still supported.|
```
nit: missing comma after suppressed

rpetrusha replied:
comma after suppressed

mairaw commented on docs/framework/migration-guide/retargeting/foreach_iterator_variable_is_now_scoped_within_the_iteration_so_closure_capturing_semantics_are_different_in_c5.md:
```diff
@@ -0,0 +1,10 @@
+
+|   |   |
+|---|---|
+|Details|Beginning with C# 5 (Visual Studio 2012), foreach iterator variables are scoped within the iteration. This can cause breaks if code was previously depending on the variables to not be included in the foreach&#39;s closure. The symptom of this change will be that an iterator variable passed to a delagate will be treated as the value it had at the time the delegate was created, rather than ...
```
inline code foreach?
will be -> try to use present tense as much as possible. it improves readability
you don't have to encode apostrophes
typo: delagate

mairaw commented on docs/framework/migration-guide/retargeting/4.0-4.5.1.md:
```diff
@@ -0,0 +1,90 @@
+
+## Core
+
+[!include[Foreach iterator variable is now scoped within the iteration, so closure capturing semantics are different (in C#5)](./foreach_iterator_variable_is_now_scoped_within_the_iteration_so_closure_capturing_semantics_are_different_in_c5.md)]
```
did you really mean to have this twice in the topic under C# and Core?

conniey replied:
In the original Breaking changes document, it is listed under both categories. I can remove one if you want. https://github.com/Microsoft/dotnet-apiport/blob/master/docs/BreakingChanges/100%20Foreach%20iterator%20variable%20is%20now%20scoped%20within%20the%20iteration.md

mairaw commented on docs/framework/migration-guide/retargeting/obsoleteattribute_exports_as_both_obsoleteattribute_and_deprecatedattribute_in_winmd_scenarios.md:
```diff
@@ -0,0 +1,10 @@
+
+|   |   |
+|---|---|
+|Details|When you create a Windows Metadata library (.winmd file), the <xref:System.ObsoleteAttribute?displayProperty=name> attribute is exported as both <xref:System.ObsoleteAttribute?displayProperty=name> and Windows.Foundation.DeprecatedAttribute.|
```
should we link to the API in docs? https://docs.microsoft.com/en-us/uwp/api/windows.foundation.metadata.deprecatedattribute

mairaw commented on docs/framework/migration-guide/retargeting/systemuri_parsing_adheres_to_rfc_3987.md:
```diff
@@ -0,0 +1,11 @@
+
+|   |   |
+|---|---|
+|Details|URI parsing has changed in several ways in .NET 4.5. Note, however, that these changes only affect code targeting .NET 4.5. If a binary targets .NET 4.0, the old behavior will be observed. Changes to URI parsing in .NET 4.5 include:<br /><ul><li>URI parsing will perform normalization and character checking according to the latest IRI rules in RFC 3987</li><li>Unicode normalization form C ...
```
the list items are missing punctuation

mairaw commented on docs/framework/migration-guide/retargeting/building_an_entity_framework_edmx_with_visual_studio_2013_can_fail_with_error_msb4062_if_using_the_entitydeploysplit_or_entityclean_tasks.md:
```diff
@@ -0,0 +1,10 @@
+|   |   |
+|---|---|
+|Details|MSBuild 12.0 tools (included in Visual Studio 2013) changed msbuild file locations, causing older Entity Framework targets files to be invalid. The result is that <code>EntityDeploySplit</code> and <code>EntityClean</code> tasks fail because they are unable to find <code>Microsoft.Data.Entity.Build.Tasks.dll</code>. Note that this break is because of a toolset (msbuild/VS) change, not be ...
+|Suggestion|Entity Framework targets files are fixed to work with the new msbuild layout beginning in the .NET Framework 4.6. Upgrading to that version of the Framework will fix this issue. Alternatively, <a href="http://stackoverflow.com/a/24249247/131944">this</a> workaround can be used to patch the targets files directly.|
```
msbuild -> MSBuild

mairaw commented on docs/framework/migration-guide/retargeting/ef_version_must_match_netfx_version.md:
```diff
@@ -0,0 +1,9 @@
+### EF version must match NetFX version
```
do customers know what netfx mean?

mairaw commented on docs/framework/migration-guide/retargeting/resolveassemblyreference_task_now_warns_of_dependencies_with_the_wrong_architecture.md:
```diff
@@ -0,0 +1,10 @@
+
+|   |   |
+|---|---|
+|Details|The task emits a warning, MSB3270, which indicates that a reference or any of its dependencies does not match the app&#39;s architecture. For example, this occurs if an app that was compiled with the anycpu option includes an x86 reference. Such a scenario could result in an app failure at run time (in this case, if the app is deployed as an x64 process).|
```
inline code anycpu?

mairaw commented on docs/framework/migration-guide/retargeting/new_ambiguous_dispatcherinvoke_overloads_could_result_in_different_behavior.md:
```diff
@@ -0,0 +1,11 @@
+
+|   |   |
+|---|---|
+|Details|The .NET Framework 4.5 adds new overloads to Dispatcher.Invoke that include a parameter of type @System.Action. When existing code is recompiled, compilers may resolve calls to Dispatcher.Invoke methods that have a @System.Delegate parameter as calls to Dispatcher.Invoke methods with an @System.Action parameter. If a call to a Dispatcher.Invoke overload with a @System.Delegate parameter  ...
```
There's an xref inside an xref link

conniey replied:
Oh boy, ahaha. :( I was doing a bunch of replaces that day... and was too aggressive about it. Lol.

mairaw commented on docs/framework/migration-guide/retargeting/some_workflow_drag_and_drop_apis_are_obsolete.md:
```diff
@@ -0,0 +1,11 @@
+### Some WorkFlow drag and drop APIs are obsolete
```
drag and drop -> drag-and-drop

mairaw commented on docs/framework/migration-guide/retargeting/workflow_30_types_are_obsolete.md:
```diff
@@ -0,0 +1,10 @@
+|   |   |
+|---|---|
+|Details|Windows Workflow Foundation (WWF) 3.0 APIs (those from the System.Workflow namespace) are now obsolete.|
+|Suggestion|New WWF 4.0 APIs (in System.Activities) should be used instead. An example of using the new APIs can be found <a href="https://docs.microsoft.com/en-us/dotnet/articles/framework/windows-workflow-foundation/how-to-update-the-definition-of-a-running-workflow-instance">here</a> and further guidance is available <a href="http://blogs.msdn.com/b/workflowteam/archive/2012/02/08/deprecatingw ...
```
here shouldn't be used as the link text and links to articles inside the repo should be relative paths

rpetrusha commented on docs/framework/migration-guide/retargeting/htmltextwriter_does_not_render_br__element_correctly.md:
```diff
@@ -0,0 +1,11 @@
+|   |   |
+|---|---|
+|Details|Beginning in the .NET Framework 4.6, calling <xref:System.Web.UI.HtmlTextWriter.RenderBeginTag(System.String)> and <xref:System.Web.UI.HtmlTextWriter.RenderEndTag> with a <code>&lt;BR /&gt;</code> element will correctly insert only one <code>&lt;BR /&gt;</code> (instead of two)|
+|Suggestion|If an app depended on the extra <code>&lt;BR /&gt;</code> tag, <xref:System.Web.UI.HtmlTextWriter.RenderBeginTag(System.String)> should be called a second time. Note that this behavior change only affects apps that are targeting the .NET Framework 4.6, so another option is to target a previous version of the .NET Framework in order to get the old behavior.|
```
This needs to be changed to "only affects apps that target the .NET Framework 4.6 or later,"

mairaw reviewed (dismissed):
a few more comments... BTW, I really like the selector on the pages!

rpetrusha commented on docs/framework/migration-guide/retargeting/throttle_concurrent_requests_per_session.md:
```diff
@@ -0,0 +1,9 @@
+
+|   |   |
+|---|---|
+|Details|In the .NET Framework 4.6.2 and earlier, Asp.net executes requests with the same Sessionid sequentially, and asp.net always issues the Sessionid through cookie by default. If a page takes long time to response, it will significantly degrade the server performance just by pressing F5 on the browser. In the fix, we added a counter to track the queued requests and terminates the requests wh ...
```
long time --> a long time
the server performance --> server performance
terminates the requests --> terminate the requests
logged in event log and HTTP 500 --> logged in the event log, and an HTTP 500
in IIS log --> in the IIS log

rpetrusha reviewed (commented):
Some comments about the individual retargeting changes.

conniey commented on docs/framework/migration-guide/retargeting/vbnet_no_longer_supports_partial_namespace_qualification_for_systemwindows_apis.md:
```diff
@@ -0,0 +1,10 @@
+
+|   |   |
+|---|---|
+|Details|Beginning in .NET 4.5.2, VB.NET projects cannot specify System.Windows APIs with partially-qualified namespaces. For example, referring to <code>Windows.Forms.DialogResult</code> will fail. Instead, code must refer to the fully qualified name (<xref:System.Windows.Forms.DialogResult>) or import the specific namespace and refer simply to <xref:System.Windows.Forms.DialogResult?displayProp ...
```
I want it rendered as such. In the original document, it appears as `DialogResult`. https://github.com/Microsoft/dotnet-apiport/blob/master/docs/BreakingChanges/127%20VB.NET%20no%20longer%20supports%20partial%20namespace%20qualification%20for%20System.Windows%20APIs.md

conniey commented on docs/framework/migration-guide/application-compatibility.md:
```diff
@@ -1,142 +1,110 @@
 ms.reviewer: ""
 ms.suite: ""
-ms.technology: 
+ms.technology:
```
That sounds like a good idea, do you want me to fill it in with "ms.technology: app-compat" in the meantime?

mairaw commented:
@conniey there are some merge conflicts here that cannot be resolved via GitHub UI. Can you check?

conniey committed:
Modify contents based on code review updates

conniey committed:
Regenerate breaking changes.

conniey commented:
@mairaw I fixed most of the warnings and the CR comments... I am having issues with one include that I had changed to be relative but it cannot be resolved. I checked the path locally and it exists. :( Should I not use an `<a>` for thsi?

The file: https://raw.githubusercontent.com/conniey/core-docs/7a6a682917c13419defd5cc1af5a4c6157e899cc/docs/includes/migration-guide/retargeting/wf/workflow-30-types-are-obsolete.md

```
Invalid file link:(~/docs/framework/windows-workflow-foundation/how-to-update-the-definition-of-a-running-workflow-instance).
```

mairaw commented:
@conniey the link needs to have the md extension in the source to make the warning go away. can you please resolve the merging conflicts again? It's not letting me do it through the UI.

conniey commented:
@mairaw I added `.md` to the link text... and rebased my commits onto master... and waiting for the new build. Is everything OK now?

conniey commented:
@mairaw I fixed the latest round of conflicting changes. The last build seemed good.

rpetrusha commented:
@conniey, there are merge conflicts again, unfortunately. Could you resolve them? 
Once there are no conflicts, is this ready to be merged?

conniey commented:
@rpetrusha @mairaw I've fixed the merge conflicts again.  This is ready to be merged.

GuardRex commented:
The build choked, so let's try again. [7:10am PST]

twsouthwick commented:
Thanks @conniey for getting this done!

rpetrusha commented:
@conniey, @twsouthwick, the build has been successful, so I'll merge. We'll promote master to live tonight, so it should be live on docs in the morning. Great work, @conniey!
